content/post/104-dht-deep-dive.md

```diff
@@ -0,0 +1,154 @@
+
+*   Provider Records: These map a data identifier (i.e., a multihash) to a peer that has advertised that they have, and are willing, to provide you with that content.
+    *   Used by IPFS to find content
+    *   Used by IPNS over PubSub to find other members of the pubsub topic
+*   IPNS Records: These map an IPNS key (i.e., hash of a public key) to an IPNS record (i.e., a signed and versioned pointer to some path like `/ipfs/bafyXYZ`)
+    *   Used by IPNS
+*   Peer Records: These map a peerID to a set of multiaddresses at which the peer may be reached
+    *   Used by IPFS when we know of a peer with content, but do not know its address
+    *   Used for manual connections (e.g., `ipfs swarm connect /p2p/QmXYZ`)
+
+Each of these record types has slightly different semantics, but they are all updated and found via the same DHT protocol, IPFS’s take on Kademlia.
+
+## Kademlia Overview
+
+The Kademlia algorithm has been around for a while and there are some great resources already available for it online, including the paper itself [link Kademlia paper on IPFS] and Wikipedia [link Kademlia wikipedia]. We’ll go through some of the basics here, though.
+
+The general idea of Kademlia is to build a DHT on top of three system parameters:
+
+*   An address space: a way that all of the peers in the network can be uniquely identified (in IPFS this is all the numbers from `0` to `2^256-1`)
+*   A metric to order the peers in the address space and therefore visualize all the peers along a line ordered from smallest to largest (in IPFS we we take `SHA256(PeerID)` and interpret it as an integer between `0` and `2^256-1`)
+*   A projection, that will take a record key and calculate a position in the address space where the peer(s) most ideally suited to store the record should be near (in IPFS we use `SHA256(Record Key)`)
+
+Having this address space and a peer ordering metric allows us to search the network as though it was a sorted list. In particular, we can turn the system into something like a skip list where a peer knows peers that are distances of around 1,2,4,8,16… away from it. This will allow us to search the list in time that is logarithmic in the size of the network (i.e. O(log(N)) lookup time). Unlike a skip list, Kademlia is somewhat unstable since peers can join, leave and rejoin the network at any time. To deal with the unstable nature of the system, a Kademlia peer does not just keep links to the peers with distance 1,2,4,8,... away from it, but instead for each multiple of 2 away keeps up to `K` (in IPFS 20) links. For example, instead of a peer keeping a single link 128 away it would keep 20 links that are between 65 and 128 away.
```

Comment:
```suggestion
Having this address space and a peer ordering metric allows us to search the network as though it was a sorted list. In particular, we can turn the system into something like a skip list where a peer knows peers that are distances of around 1,2,4,8,16… away from it. This will allow us to search the list in time that is logarithmic in the size of the network (i.e. O(log(N)) lookup time). Unlike a skip list, Kademlia is somewhat unstable since peers can join, leave and rejoin the network at any time. To deal with the unstable nature of the system, a Kademlia peer does not just keep links to the peers with distance 1,2,4,8,... away from it, but instead for each multiple of 2 away it keeps up to `K` (in IPFS 20) links. For example, instead of a peer keeping a single link 128 away it would keep 20 links that are between 65 and 128 away.
```